include/swift/Runtime/Mutex.h

```diff
@@ -0,0 +1,318 @@
+//===--- Mutex.h - Lockables ------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Mutex, Condition, and Scoped lock abstactions for use in Swift runtime.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_RUNTIME_MUTEX_H
+#define SWIFT_RUNTIME_MUTEX_H
+
+#if (defined(__APPLE__) || defined(__linux__) || defined(__CYGWIN__))
+#define SWIFT_RUNTIME_MUTEX_HAVE_PHTREADS
+#endif
+
+#ifdef SWIFT_RUNTIME_MUTEX_HAVE_PHTREADS
+#include <pthread.h>
+#endif
```

Comment:
We should have an `#else #error` branch here so that people porting to non-pthread platforms know they have some work to do here.